File: plugins/unityshell/src/ShortcutHint.cpp
// -*- Mode: C++; indent-tabs-mode: nil; tab-width: 2 -*-
/*
 * Copyright (C) 2011 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authored by: Andrea Azzarone <azzaronea@gmail.com>
 */

#include "ShortcutHint.h"

#include <core/core.h> // Compiz...
#include <NuxCore/Logger.h>

namespace unity
{
namespace shortcut
{
namespace
{
  nux::logging::Logger logger("unity.shortcut");
} // anonymouse namespace 

// Ctor
Hint::Hint(std::string category, std::string prefix,
           std::string postfix,std::string description,
           OptionType type, std::string arg1, 
           std::string arg2, std::string arg3)
{
  Category = category;
  Prefix = prefix;
  Postfix = postfix;
  Description = description;
  Type = type;
  Arg1 = arg1;
  Arg2 = arg2;
  Arg3 = arg3;
}

// Dtor
Hint::~Hint()
{
}

// Copy Ctor
Hint::Hint(unity::shortcut::Hint const& obj)
{
  Category = obj.Category();
  Prefix = obj.Prefix();
  Postfix = obj.Postfix();
  Description = obj.Description();
  Type = obj.Type();
  Arg1 = obj.Arg1();
  Arg2 = obj.Arg2();
  Arg3 = obj.Arg3();
  Value = obj.Value();
}


/*
 * Gets and fills the shortcut value. 
 * Returns true if everything was OK, returns false otherwise.
 * Use member property Value to get it.
 */
bool Hint::Fill()
{
  // Reset Value.
  Value = "";

  switch(Type())
  {
    case COMPIZ_OPTION:
    {
      // Arg1 = Plugin name
      // Arg2 = Option name
      CompPlugin* p = CompPlugin::find(Arg1().c_str());

      if (!p) 
        return false;

      foreach (CompOption &opt, p->vTable->getOptions())
      {
          if (opt.name() == Arg2())
          {
            Value = opt.value().action().keyToString();
            return true;
          }
      }

      break;
    }
    case HARDCODED_OPTION:
      Value = Arg1();
      return true;

    default:
      LOG_WARNING(logger) << "Unable to find the option type" << Type(); 
  }

  return false;
}

} // namespace shortcut
} // namespace unity

File: tests/autopilot/autopilot/emulators/unity/shortcut_hint.py
# -*- Mode: Python; coding: utf-8; indent-tabs-mode: nil; tab-width: 4 -*-
# Copyright 2012 Canonical
# Author: Marco Trevisan (Treviño)
#
# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 3, as published
# by the Free Software Foundation.
#

import logging

from autopilot.emulators.unity import UnityIntrospectionObject
from autopilot.keybindings import KeybindingsHelper


logger = logging.getLogger(__name__)


class ShortcutController(UnityIntrospectionObject, KeybindingsHelper):
    """ShortcutController proxy class."""

    def show(self):
        """Push the keys necessary to reveal the shortcut hint, but don't block."""
        logger.debug("Revealing shortcut hint with keyboard.")
        self.keybinding_hold("shortcuthint/reveal")

    def ensure_visible(self):
        """Block until the shortcut hint is visible."""
        if not self.visible:
            logger.debug("Revealing shortcut hint with keyboard.")
            self.keybinding_hold("shortcuthint/reveal")
            self.visible.wait_for(True)

    def hide(self):
        """Release the keys keeping the shortcut hint open, but don't block."""
        logger.debug("Un-revealing shortcut hint with keyboard.")
        self.keybinding_release("shortcuthint/reveal")

    def ensure_hidden(self):
        """Block until the shortcut hint is hidden."""
        if self.visible:
            logger.debug("Un-revealing shortcut hint with keyboard.")
            self.keybinding_release("shortcuthint/reveal")
            self.visible.wait_for(False)

    def cancel(self):
        logger.debug("Hide the shortcut hint with keyboard, without releasing the reveal key.")
        self.keybinding("shortcuthint/cancel")

    def get_show_timeout(self):
        return self.timeout_duration / 1000.0
